compiler/optimizer/LoopVersioner.cpp

```diff
@@ -7946,9 +7965,144 @@ void TR_LoopVersioner::collectAllExpressionsToBeChecked(List<TR::TreeTop> *nullC
       }
    }
 
+/**
+ * \brief Determine whether \p node has a stable value when re-evaluated, or
+ * whether privatization is needed to force the value to remain stable.
+ *
+ * Only the evaluation of \p node itself is considered here. Its children (and
+ * generally descendants) are assumed to have stable values, possibly because
+ * they themselves will be privatized.
+ *
+ * **For debugging purposes only**, the \c TR_nothingRequiresPrivatizationInVersioner
+ * environment variable can be set to cause this method to assume that
+ * privatization is unnecessary.
+ *
+ * \param node The node in question
+ * \return true if privatization is required
+ *
+ * \see suppressInvarianceAndPrivatization()
+ */
+bool TR_LoopVersioner::requiresPrivatization(TR::Node *node)
+   {
+   static const bool nothingRequiresPrivatization =
+      feGetEnv("TR_nothingRequiresPrivatizationInVersioner") != NULL;
+
+   if (nothingRequiresPrivatization)
+      return false;
+
+   if (!node->getOpCode().hasSymbolReference())
+      return false;
+
+   if (node->getOpCodeValue() == TR::loadaddr || node->getOpCode().isTreeTop())
+      return false;
+
+   TR::SymbolReference *symRef = node->getSymbolReference();
+   TR::Symbol *sym = symRef->getSymbol();
+   if (sym->isAutoOrParm())
+      return false;
+
+   TR::SymbolReferenceTable *srTab = comp()->getSymRefTab();
+   TR::ResolvedMethodSymbol *curMethod = comp()->getMethodSymbol();
+   if (symRef == srTab->findOrCreateInstanceOfSymbolRef(curMethod))
+      return false;
+
+   switch (symRef->getReferenceNumber() - srTab->getNumHelperSymbols())
+      {
+      case TR::SymbolReferenceTable::vftSymbol:
+      case TR::SymbolReferenceTable::javaLangClassFromClassSymbol:
+      case TR::SymbolReferenceTable::classFromJavaLangClassSymbol:
+         return false;
+
+      default:
+         break;
+      }
+
+   if (srTab->isVtableEntrySymbolRef(symRef))
+      return false;
+
+   if (suppressInvarianceAndPrivatization(symRef))
+      return false;
+
+   return true;
+   }
+
+/**
+ * \brief Determine whether to prevent accesses to \p symRef from being
+ * considered invariant and getting privatized, even if it doesn't appear to be
+ * modified in the loop.
+ *
+ * This is useful for values that are not explicitly written in IL but may be
+ * set by the runtime system, e.g. the current exception, and for values for
+ * which stores from other threads should be observed (without imposing any
+ * other memory ordering effect).
+ *
+ * \param symRef The symbol reference for the operation in question
+ *
+ * \return true if node should be considered non-invariant and shouldn't be
+ * privatized
+ *
+ * \see isExprInvariant()
+ * \see requiresPrivatization()
+ */
+bool TR_LoopVersioner::suppressInvarianceAndPrivatization(TR::SymbolReference *symRef)
```

Comment:
Should we have an api like this generalized to a code location where other optimizations that have a notion of invariance, e.g. PRE and local CSE can also call ?

Reply:
Quite possibly - I recently observed PRE hoisting a load of `j9methodConstantPoolSymbol`. However, for the sake of this fix I'd like to leave this logic here, to be moved afterward if desired, if that's okay with you.

Reply:
Okay can an issue be opened please (to move the logic) ?

Reply:
Opened #4140